gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc

```diff
@@ -139,38 +190,90 @@ int main(int argc, char* argv[]) {
+      if (addrs->accessed_blocks.size() > 0) {
+        output_file << kMemDefPrefix << kMemNamePrefix << " "
+                    << addrs->block_size << " " << initial_mem_val_str << "\n";
+      }
+      for (const auto& addr : addrs->accessed_blocks) {
+        output_file << kMemMapPrefix << kMemNamePrefix << " " << std::dec
+                    << addr << "\n";
+      }
+
+      // Append disassembled instructions.
+      for (const auto& instr : proto->machine_instructions()) {
+        output_file << instr.assembly() << "\n";
+      }
     }
 
-    // Append disassembled instructions.
-    for (const auto& instr : proto->machine_instructions()) {
-      output_file << instr.assembly() << "\n";
+    if (json_output_enabled) {
+      llvm::json::Object current_snippet;
+
+      if (addrs->accessed_blocks.size() > 0) {
+        llvm::json::Array memory_definitions;
+        llvm::json::Object current_memory_definition;
+        current_memory_definition["Name"] = llvm::json::Value(kMemNamePrefix);
+        current_memory_definition["Size"] =
+            llvm::json::Value(addrs->block_size);
+        current_memory_definition["Value"] = llvm::json::Value(kInitialMemVal);
+        memory_definitions.push_back(std::move(current_memory_definition));
+        current_snippet["MemoryDefinitions"] =
+            llvm::json::Value(std::move(memory_definitions));
+
+        llvm::json::Array memory_mappings;
+        for (const uintptr_t addr : addrs->accessed_blocks) {
+          llvm::json::Object current_memory_mapping;
+          current_memory_mapping["Value"] = llvm::json::Value(kMemNamePrefix);
+          current_memory_mapping["Address"] = llvm::json::Value(addr);
+          memory_mappings.push_back(std::move(current_memory_mapping));
+        }
+        current_snippet["MemoryMappings"] =
+            llvm::json::Value(std::move(memory_mappings));
+      } else {
+        current_snippet["MemoryDefinitions"] = llvm::json::Array();
+        current_snippet["MemoryMappings"] = llvm::json::Array();
+      }
+
+      std::string hex_string = {hex.begin(), hex.end()};
+      current_snippet["Hex"] = llvm::json::Value(hex_string);
+
+      processed_snippets.push_back(
+          llvm::json::Value(std::move(current_snippet)));
+
+      if (file_counter % json_split_count == 0) {
+        size_t json_file_number = file_counter / json_split_count;
+        bool write_successfully = write_json_file(
+            std::move(processed_snippets), json_file_number, json_output_dir);
+        if (!write_successfully) return 4;
+        processed_snippets.clear();
+      }
     }
 
     file_counter++;
   }
-}
\ No newline at end of file
+
+  if (json_output_enabled) {
+    size_t json_file_number = file_counter / json_split_count;
+    bool write_successfully = write_json_file(
+        std::move(processed_snippets), json_file_number, json_output_dir);
+    if (!write_successfully) return 4;
+  }
+}
```

Comment:
We're missing `return 0;` at the end. It was missing even before, but since the file is already changed in this PR...